staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -947,31 +1035,56 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer, budget *ti
 	budget.returnUnused(timeout - time.Since(startTime))
 }
 
-// NOTE: sendWatchCacheEvent is assumed to not modify <event> !!!
-func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
-	curObjPasses := event.Type != watch.Deleted && c.filter(event.Key, event.ObjLabels, event.ObjFields)
-	oldObjPasses := false
-	if event.PrevObject != nil {
-		oldObjPasses = c.filter(event.Key, event.PrevObjLabels, event.PrevObjFields)
-	}
-	if !curObjPasses && !oldObjPasses {
-		// Watcher is not interested in that object.
-		return
+func (c *cacheWatcher) addBookmark(event *watchCacheEvent) {
+	// send the bookmark event when the client support it
+	if c.heartbeatPeriod() > 0 {
+		// If we can't send it, don't block on it.
+		select {
+		case c.input <- event:
+		default:
+		}
 	}
+}
+
+// It seems that sending bookmark to every watcher every X seconds is too expensive,
```

Comment:
we can make X large enough. I feel every minute should be OK (short than the expected etcd compaction duration). We also should send the bookmark if we know the watcher is about to close.

Reply:
Agree on sending Bookmark every minute should be little impact on performance. We can do that by find a number n which (Timeout-2)/n > 1 minute > (Timeout-2)/(n+1).